nmdm: Allow module to be force-unloaded.

MegaManSec opened the pull request:
If an nmdm device was created by accident such as by running
`ls /dev/nmdmAA`, it would restrict the module from being unloaded,
despite no data passing through it.

With this patch, `kldunload nmdm` will still fail if any nmdm device still
exists.

`kldunload -f nmdm` will try to destroy all nmdm devices before unloading.
If an nmdm modem is either opened in blocking mode or data is being
passed through it, after five seconds of trying, it will not be destroyed
and the unload will also fail.

Likewise, fix an incorrect tty_lock, a typo, and the order of locking.

MegaManSec committed:
nmdm: Fix double-lock of tp in nmdm.

Also fix the order (which is not important).
Also fix a typo.

Signed-off-by: Joshua Rogers <Joshua@Joshua.Hu>

MegaManSec committed:
nmdm: Allow module to be force-unloaded.

If an nmdm device was created by accident such as by running
`ls /dev/nmdmAA`, it would restrict the module from being unloaded,
despite no data passing through it.

With this patch, `kldunload nmdm` will still fail if any nmdm device still
exists.

`kldunload -f nmdm` will try to destroy all nmdm devices before unloading.
If an nmdm modem is either opened in blocking mode or data is being
passed through it, after five seconds of trying, it will not be destroyed
and the unload will also fail.

Signed-off-by: Joshua Rogers <Joshua@Joshua.Hu>

bsdimp reviewed (approved):
This looks like a good fix to me. @kevans91 is the closest thing we have to a tty expert though.

kevans91 commented on sys/dev/nmdm/nmdm.c:
```diff
@@ -115,16 +115,16 @@ nmdm_close(struct tty *tp)
 		return;
 
 	/* Shut down self. */
+	tty_lock(tp);
```
This is wrong, `tp` is locked on entry

MegaManSec replied:
Right: I mixed up the locks here, and this effective double locking (in the later commit where this lock is called from nmdm_destroy) hides a bug too. I'll make some changes and post them here.

MegaManSec replied:
"Fixed" within the realm of the second and third patch.

kevans91 commented on sys/dev/nmdm/nmdm.c:
```diff
@@ -115,16 +115,16 @@ nmdm_close(struct tty *tp)
 	tty_rel_gone(tp);
 
 	/* Shut down second part. */
-	tty_lock(tp);
```
This is necessary because they share a single mutex, and it needs to be held for the below `tty_rel_gone`.  I'd be in favor of switching this one to `otp` to make it more clear, though.

MegaManSec replied:
"Fixed" within the realm of the second and third patch.

kevans91 commented on sys/dev/nmdm/nmdm.c:
```diff
@@ -115,16 +115,16 @@ nmdm_close(struct tty *tp)
 	otp = onp->np_tty;
+	tty_lock(otp);
 	tty_rel_gone(otp);
-	tty_lock(tp);
```
This needs to be restored because this tty needs to remain locked on exit, but we dropped it with `tty_rel_gone` again

MegaManSec replied:
"Fixed" within the realm of the second and third patch.

kevans91 reviewed (changes requested):
Some problems with the locking bits, and IMO the other half is working around a bug.  I don't see where the current implementation would be intentionally stopping destruction of a nmdm pair that was only ever half open.

kevans91 commented:
> Some problems with the locking bits, and IMO the other half is working around a bug. I don't see where the current implementation would be intentionally stopping destruction of a nmdm pair that was only ever half open.

Ah, right- we clone it in response to stat(), nothing actually opened it or ever planned on opening it. Hmm.

kevans91 commented on sys/dev/nmdm/nmdm.c:
```diff
@@ -416,6 +452,29 @@ nmdm_outwakeup(struct tty *tp)
+{
+	struct nmdmsoftc *ns;
+	sbintime_t start_time = sbinuptime();
+	sbintime_t timeout = SBT_1S * 5; // Try to force unload for 5 seconds
```
I think this is entirely valid, but I'd really prefer to accomplish it without the timeout by adding some plumbing through the tty later to drain the deferred destroy_dev taskqueue. kern_conf.c doesn't seem to have quite the KPI that I think I'd like to see.

Ideally, I think we'd have some kind of `tty_drain_freed` -> `destroy_dev_drain_pending` -> `taskqueue_drain` that we can call to force the async bits to run, then we can just EBUSY it if we still have nmdm devices when we get back.

kevans91 commented on sys/dev/nmdm/nmdm.c:
```diff
@@ -416,6 +452,29 @@ nmdm_outwakeup(struct tty *tp)
+
+	mtx_lock(&nmdmsoftc_lock);
+	TAILQ_FOREACH(ns, &nmdmsoftc_list, ns_list) {
+		nmdm_destroy(ns->ns_part1.np_tty);
```
... and `nmdm_destroy` should ideally return EBUSY if one of them has both ttys and at least one of the ttys is opened, so that we don't bother waiting at all if we already know it's not going to magically go away.

kevans91 reviewed (commented):
Actionable feedback on a path forward for the other half...

MegaManSec committed:
Ensure each tty is correctly locked and unlocked on entry and unload.

On entry to nmdm_close(), the tty is already locked. nmdm_destroy()
does not return before the tty is unlocked by tty_rel_gone().
This means that  we need to lock the tty when the nmdm_close()
returns.

nmdm_destroy() assumes that the tty is already locked. Therefore,
nmdm_unload() must lock the tty before calling nmdm_destroy().

Signed-off-by: Joshua Rogers <Joshua@Joshua.Hu>

bsdimp commented:
@kevans91 Looks like this is ready for you to re-review since it looks like your feedback has been applied.

MegaManSec commented:
The remaining recommendation "_Ideally, I think we'd have some kind of tty_drain_freed -> destroy_dev_drain_pending -> taskqueue_drain_" is not done yet: I have not had time yet to try this.

bsdimp commented:
any update?

bsdimp commented:
last pinged over 6 months ago, with no update. Closing.
Feel free to re-open this or to create a new pull request if the required changes are made. Thanks for your submission. Sorry that it didn't make it in. But we're open to having a fixed version.

MegaManSec commented:
Unfortunately, I never got time to make the changes requested ("without the timeout by adding some plumbing through the tty layer to drain the deferred destroy_dev taskqueue").

Perhaps a freebsd bug report should be made for this, though.

MegaManSec commented:
addressed in https://github.com/freebsd/freebsd-src/pull/2207
